builtin/credential/aws/path_login.go

```diff
@@ -0,0 +1,483 @@
+	if givenPendingTime.Before(storedPendingTime) {
+		return fmt.Errorf("instance meta-data is older than the one used for previous login")
+	}
+	return nil
+}
+
+// Verifies the correctness of the authenticated attributes present in the PKCS#7
+// signature. After verification, extracts the instance identity document from the
+// signature, parses it and returns it.
+func parseIdentityDocument(s logical.Storage, pkcs7B64 string) (*identityDocument, error) {
+	pkcs7B64 = fmt.Sprintf("-----BEGIN PKCS7-----\n%s\n-----END PKCS7-----", pkcs7B64)
+
+	// Decode the PEM encoded signature.
+	pkcs7BER, pkcs7Rest := pem.Decode([]byte(pkcs7B64))
+	if len(pkcs7Rest) != 0 {
+		return nil, fmt.Errorf("failed to decode the PEM encoded PKCS#7 signature")
+	}
+
+	// Parse the signature from asn1 format into a struct.
+	pkcs7Data, err := pkcs7.Parse(pkcs7BER.Bytes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the BER encoded PKCS#7 signature: %s\n", err)
+	}
+
+	// Get the public certificate that is used to verify the signature.
+	publicCert, err := awsPublicCertificateParsed(s)
+	if err != nil {
+		return nil, err
+	}
+	if publicCert == nil {
+		return nil, fmt.Errorf("certificate to verify the signature is not found")
+	}
+
+	// Before calling Verify() on the PKCS#7 struct, set the certificate to be used
+	// to verify the contents in the signer information.
+	pkcs7Data.Certificates = []*x509.Certificate{publicCert}
+
+	// Verify extracts the authenticated attributes in the PKCS#7 signature, and verifies
+	// the authenticity of the content using 'dsa.PublicKey' embedded in the public certificate.
+	if pkcs7Data.Verify() != nil {
+		return nil, fmt.Errorf("failed to verify the signature")
+	}
+
+	// Check if the signature has content inside of it.
+	if len(pkcs7Data.Content) == 0 {
+		return nil, fmt.Errorf("instance identity document could not be found in the signature")
+	}
+
+	var identityDoc identityDocument
+	err = json.Unmarshal(pkcs7Data.Content, &identityDoc)
+	if err != nil {
+		return nil, err
+	}
+
+	return &identityDoc, nil
+}
+
+// pathLoginUpdate is used to create a Vault token by the EC2 instances
+// by providing its instance identity document, pkcs7 signature of the document,
+// and a client created nonce.
+func (b *backend) pathLoginUpdate(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	pkcs7B64 := data.Get("pkcs7").(string)
+
+	if pkcs7B64 == "" {
+		return logical.ErrorResponse("missing pkcs7"), nil
+	}
+
+	// Verify the signature of the identity document.
+	identityDoc, err := parseIdentityDocument(req.Storage, pkcs7B64)
+	if err != nil {
+		return nil, err
+	}
+	if identityDoc == nil {
+		return logical.ErrorResponse("failed to extract instance identity document from PKCS#7 signature"), nil
+	}
+
+	clientNonce := data.Get("nonce").(string)
+	if clientNonce == "" {
+		return logical.ErrorResponse("missing nonce"), nil
+	}
+
+	// Allowing the lengh of UUID for a client nonce.
```

Comment:
We should probably allow more than this in case people really want to use longer nonces. A 256-bit nonce in hex will be 64 characters, so let's make that the max.